README.md
# Image Upsampling

This Python script performs image upsampling using the OpenCV library. It takes an input image, upsamples it by a specified scale factor, and saves the upsampled image to disk.

## Prerequisites

- Python 3.x
- OpenCV library (`cv2`)

## Usage

1. Ensure that you have the necessary prerequisites installed.

2. Download the script and save it to your desired location.

3. Place the input image you want to upsample in the same directory as the script and name it "2019me.jpg" (or modify the code to specify the correct filename).

4. Open a terminal or command prompt and navigate to the directory containing the script.

5. Run the script by executing the following command:

    ```bash
    python script_name.py
    ```

   Make sure to replace `script_name.py` with the actual name of the script file.

6. The script will read the input image, upsample it using the Lanczos interpolation algorithm, and save the upsampled image as "upsampled2019.jpg" in the same directory.

## Function

### `upsample_image(image, scale)`

This function takes an image and a scale factor as input and returns the upsampled image. It uses the following steps to upsample the image:

1. Get the dimensions of the input image.
2. Calculate the new dimensions based on the scale factor.
3. Upsample the image using the Lanczos interpolation algorithm with the `cv2.resize()` function.
4. Return the upsampled image.

### `main()`

The `main()` function serves as the entry point of the script. It performs the following tasks:

1. Reads the input image from disk.
2. Calls the `upsample_image()` function to upsample the image with a scale factor of 4.
3. Saves the upsampled image to disk as "upsampled2019.jpg".

## Notes

- The script uses the Lanczos interpolation algorithm (`cv2.INTER_LANCZOS4`) for image upsampling. You can modify this if you prefer a different interpolation algorithm by changing the `interpolation` parameter in the `cv2.resize()` function.
- Ensure that the input image is in the same directory as the script and has the correct filename specified in the code.
- The script overwrites any existing "upsampled2019.jpg" file in the directory, so be cautious if you have a file with the same name.
- You can modify the code to suit your specific image and scaling requirements.

Feel free to explore and modify the code to enhance the image upsampling functionality according to your needs.


pichance/CHANGELOG
# Pic Hance Release Notes

## 0.0.1 (05 Jun 2023)

* Initial release


imgUpscale.py
#!/usr/bin/python3

import cv2

def upsample_image(image, scale):
    """Upsamples an image by the given scale factor.

    Args:
        image: The image to upsample.
        scale: The scale factor to upsample by.

    Returns:
        The upsampled image.
    """

    # Get the dimensions of the image.
    height, width = image.shape[:2]

    # Calculate the new dimensions based on the scale factor.
    new_height = int(height * scale)
    new_width = int(width * scale)

    # Upsample the image using the Lanczos interpolation algorithm.
    upsampled_image = cv2.resize(image, (new_width, new_height), interpolation=cv2.INTER_LANCZOS4)

    # Return the upsampled image.
    return upsampled_image

def main():
    # Read the image from disk.
    image = cv2.imread("2019me.jpg")
 
    # Upsample the image to 4K.
    upsampled_image = upsample_image(image, 4)

    # Save the upsampled image to disk.
    cv2.imwrite("upsampled2019.jpg", upsampled_image)

if __name__ == "__main__":
    main()


pichance/src/pichance/pichance.py
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, CENTER
from toga.widgets import FileSelection

import cv2
import urllib.request

def upsample_image(image, scale):
    # Upsample the image using the Lanczos interpolation algorithm.
    upsampled_image = cv2.resize(image, None, fx=scale, fy=scale, interpolation=cv2.INTER_LANCZOS4)
    return upsampled_image

def upscale_image(widget):
    # Check if an image file has been selected.
    if widget.file_selection.filename:
        # Read the image from disk.
        image = cv2.imread(widget.file_selection.filename)

        # Upscale the image to 4x.
        scale_factor = 4
        upsampled_image = upsample_image(image, scale_factor)

        # Save the upsampled image to disk.
        output_filename = widget.file_selection.filename + "_upscaled.jpg"
        cv2.imwrite(output_filename, upsampled_image)

        # Update the label to display a success message.
        widget.label.text = "Image upscaled and ready for download!"

        # Set the download URL for the upscaled image.
        widget.download_button.enabled = True
        widget.download_button.url = urllib.request.pathname2url(output_filename)

def main():
    # Create the main application window.
    app = toga.App("Image Upscaler", "com.example.imageupscaler")

    # Create a file selection widget for image upload.
    file_selection = FileSelection()

    # Create a button widget for upscaling the image.
    upscale_button = toga.Button('Upscale', on_press=upscale_image)

    # Create a label widget to display messages.
    label = toga.Label('Upload an image and press the "Upscale" button.')

    # Create a download button widget for the upscaled image.
    download_button = toga.Button('Download', enabled=False)

    # Create a box to contain the widgets.
    box = toga.Box(children=[file_selection, upscale_button, label, download_button], style=Pack(direction=COLUMN, padding=20))

    # Define the download button behavior.
    def on_download_button_press(widget):
        # Open the upscaled image file.
        with open(widget.url, 'rb') as file:
            content = file.read()
        # Prompt the user to save the file.
        save_file_dialog = toga.save_file_dialog(file_extension='.jpg')
        if save_file_dialog.show_modal():
            save_file_path = save_file_dialog.path
            with open(save_file_path, 'wb') as save_file:
                save_file.write(content)

    # Set the download button behavior.
    download_button.on_press = on_download_button_press

    # Create a main window with the box as its content.
    main_window = toga.MainWindow(title="Image Upscaler", size=(400, 300))
    main_window.content = box

    # Start the application event loop.
    app.main_window = main_window
    app.run()

if __name__ == "__main__":
    main()


pichance/src/pichance/app.py
import toga
from toga.style import Pack
from toga.style.pack import COLUMN
import cv2
import urllib.request
from tkinter import filedialog
import os

class PicHance(toga.App):
    def startup(self):
        main_box = toga.Box(style=Pack(direction=COLUMN, padding=10))

        # Create a button widget for image upload.
        upload_button = toga.Button('Upload Image', on_press=self.upload_image)

        # Create a button widget for upscaling the image.
        upscale_button = toga.Button('Upscale', on_press=self.upscale_image, enabled=False)

        # Create a label widget to display messages.
        self.label = toga.Label('Upload an image and press the "Upscale" button.')

        # Create a download button widget for the upscaled image.
        self.download_button = toga.Button('Download', on_press=self.download_image, enabled=False)

        # Add widgets to the main box
        main_box.add(upload_button)
        main_box.add(upscale_button)
        main_box.add(self.label)
        main_box.add(self.download_button)

        # Create a main window
        self.main_window = toga.MainWindow(title=self.formal_name, size=(400, 300))
        self.main_window.content = main_box
        self.main_window.show()

    def upload_image(self, widget):
        # Open a file dialog for image selection.
        file_path = filedialog.askopenfilename(filetypes=[("Image Files", "*.jpg;*.jpeg;*.png")])
        if file_path:
            self.image_path = file_path
            self.label.text = f"Image selected: {self.image_path}"
            self.download_button.enabled = False

    def upscale_image(self, widget):
        # Read the image from disk.
        image = cv2.imread(self.image_path)

        # Upscale the image to 4x.
        scale_factor = 4
        upsampled_image = cv2.resize(image, None, fx=scale_factor, fy=scale_factor, interpolation=cv2.INTER_LANCZOS4)

        # Save the upsampled image to disk.
        output_filename = self.image_path + "_upscaled.jpg"
        cv2.imwrite(output_filename, upsampled_image)

        # Update the label to display a success message.
        self.label.text = "Image upscaled and ready for download!"

        # Set the download path for the upscaled image.
        self.download_path = output_filename
        self.download_button.enabled = True

    def download_image(self, widget):
        # Prompt the user to save the file.
        save_file_path = filedialog.asksaveasfilename(defaultextension=".jpg", filetypes=[("JPEG Image", "*.jpg")])
        if save_file_path:
            # Copy the upscaled image to the desired save path.
            os.replace(self.download_path, save_file_path)
            self.label.text = "Image downloaded successfully!"

    def on_close(self):
        # Ensure the temporary file is deleted when the app is closed
        if hasattr(self, 'download_path') and os.path.exists(self.download_path):
            os.remove(self.download_path)

def main():
    return PicHance()
